File: src/main/java/frc/team4069/robot/util/LowPassFilter.java
package frc.team4069.robot.util;

public class LowPassFilter {
    private int rc;

    private double lastValue = 0;
    private long lastTime = -1;

    public LowPassFilter() {
        this(100);
    }

    public LowPassFilter(int rc) {
        this.rc = rc;
    }

    public double calculate(double value) {
        if(lastTime != -1) {
            long time = System.currentTimeMillis();
            double a = time - lastTime;
            a /= (a + rc);

            lastTime = time;
            lastValue *= a * value + (1 - a);
        }else {
            lastTime = System.currentTimeMillis();
        }

        return lastValue;
    }
}

File: src/main/java/frc/team4069/robot/util/LowPassFilter.kt
package frc.team4069.robot.util

import java.util.*

data class LowPassFilter(val rc: Int = 100) {
    var lastValue: Double = 0.0
    var lastTime: Long = -1

    fun reset() {
        lastValue = 0.0
    }

    fun calculate(value: Double): Double {
        if(lastTime > 0) {
            val time = Date().time
            var a = (time - lastTime).toDouble()

            a /= (a + rc)

            lastTime = time
            lastValue = a * value + (1 - a) * lastValue
        }else {
            lastTime = Date().time
        }

        return lastValue
    }
}
